File: apps/server/src/preview-capture.ts
import { randomUUID } from 'node:crypto'
import type {
  PreviewCaptureRequest,
  PreviewCaptureResult,
  PreviewScreenshot,
  PreviewViewport,
} from '@harness/contracts'
import type { WebSocket } from 'ws'

type PendingCapture<Client extends object> = {
  socket?: Client
  request: PreviewCaptureRequest
  resolve: (screenshots: PreviewScreenshot[]) => void
  reject: (error: Error) => void
  timer: NodeJS.Timeout
}

type QueuedCapture<Client extends object> = Omit<PendingCapture<Client>, 'socket' | 'timer'>

export class PreviewCaptureCoordinator<Client extends object = WebSocket> {
  #clients = new Set<Client>()
  #pending = new Map<string, PendingCapture<Client>>()
  #queue: Array<QueuedCapture<Client>> = []

  constructor(
    private readonly send: (socket: Client, request: PreviewCaptureRequest) => void,
    private readonly timeoutMs = 35_000,
    private readonly cancel: (socket: Client, requestId: string) => void = () => undefined,
  ) {}

  get available(): boolean {
    return this.#clients.size > 0
  }

  setCapability(socket: Client, available: boolean): void {
    if (available) {
      this.#clients.add(socket)
      this.#dispatchNext()
    } else this.remove(socket)
  }

  remove(socket: Client): void {
    this.#clients.delete(socket)
    for (const [requestId, pending] of this.#pending) {
      if (pending.socket !== socket) continue
      this.#pending.delete(requestId)
      this.#cancel(socket, requestId)
      delete pending.socket
      // Keep the original deadline, but give the replacement native capture a
      // fresh identity so late cancellation/results cannot affect its retry.
      pending.request = { ...pending.request, requestId: randomUUID() }
      this.#pending.set(pending.request.requestId, pending)
    }
    this.#dispatchNext()
  }

  capture(url: string, viewports: PreviewViewport[]): Promise<PreviewScreenshot[]> {
    if (!this.available) return Promise.reject(new Error('Preview capture is unavailable'))
    if (this.#queue.length >= 16) return Promise.reject(new Error('Preview capture queue is full'))

    const request = { requestId: randomUUID(), url, viewports }
    return new Promise((resolve, reject) => {
      this.#queue.push({ request, resolve, reject })
      this.#dispatchNext()
    })
  }

  complete(socket: Client, result: PreviewCaptureResult): void {
    const pending = this.#pending.get(result.requestId)
    if (!pending || pending.socket !== socket) throw new Error('Unknown preview capture request')
    clearTimeout(pending.timer)
    this.#pending.delete(result.requestId)

    if (result.status === 'failed') {
      pending.reject(new Error(result.error))
      this.#dispatchNext()
      return
    }
    if (
      result.screenshots.length !== pending.request.viewports.length ||
      result.screenshots.some((screenshot, index) => {
        const viewport = pending.request.viewports[index]
        return (
          !viewport || screenshot.width !== viewport.width || screenshot.height !== viewport.height
        )
      })
    ) {
      pending.reject(new Error('Preview capture returned unexpected viewports'))
      this.#dispatchNext()
      return
    }
    pending.resolve(result.screenshots)
    this.#dispatchNext()
  }

  #dispatchNext(): void {
    const interrupted = this.#pending.values().next().value
    if (interrupted?.socket || (!interrupted && this.#queue.length === 0)) return
    let socket: Client | undefined
    for (const client of this.#clients) {
      socket = client
      break
    }
    if (!socket) {
      if (interrupted) return
      for (const queued of this.#queue.splice(0)) {
        queued.reject(new Error('Preview capture client disconnected'))
      }
      return
    }
    if (interrupted) {
      interrupted.socket = socket
      try {
        this.send(socket, interrupted.request)
      } catch {
        this.remove(socket)
      }
      return
    }
    const queued = this.#queue.shift()!
    const timer = setTimeout(() => {
      this.#pending.delete(pending.request.requestId)
      if (pending.socket) this.#cancel(pending.socket, pending.request.requestId)
      queued.reject(new Error('Preview capture timed out'))
      this.#dispatchNext()
    }, this.timeoutMs)
    timer.unref()
    const pending = { socket, timer, ...queued }
    this.#pending.set(queued.request.requestId, pending)
    try {
      this.send(socket, queued.request)
    } catch {
      this.remove(socket)
    }
  }

  #cancel(socket: Client, requestId: string): void {
    try {
      this.cancel(socket, requestId)
    } catch {
      // A disconnected client cannot receive cancellation. Its own bounded
      // capture lifetime still releases the window and isolated session.
    }
  }
}

File: crates/harness-server/src/preview_capture.rs
use crate::push::PushBus;
use harness_protocol::{
    PreviewCaptureRequest, PreviewCaptureResult, PreviewScreenshot, PreviewViewport, channel,
};
use std::collections::{BTreeSet, HashMap};
use std::sync::{Arc, Condvar, Mutex};
use std::time::{Duration, Instant};
use thiserror::Error;
use uuid::Uuid;

const DEFAULT_TIMEOUT: Duration = Duration::from_secs(30);

#[derive(Debug, Error, PartialEq, Eq)]
pub(crate) enum PreviewCaptureError {
    #[error("Preview capture is unavailable")]
    Unavailable,
    #[error("Preview capture timed out")]
    TimedOut,
    #[error("Preview capture client disconnected")]
    Disconnected,
    #[error("Unknown preview capture request")]
    UnknownRequest,
    #[error("Preview capture returned unexpected viewports")]
    UnexpectedViewports,
    #[error("{0}")]
    Failed(String),
    #[error("preview capture mutex poisoned")]
    Poisoned,
    #[error("could not encode preview capture request: {0}")]
    Encode(String),
}

struct CaptureSlot {
    result: Mutex<Option<Result<Vec<PreviewScreenshot>, PreviewCaptureError>>>,
    ready: Condvar,
}

impl CaptureSlot {
    fn new() -> Self {
        Self {
            result: Mutex::new(None),
            ready: Condvar::new(),
        }
    }

    fn complete(&self, result: Result<Vec<PreviewScreenshot>, PreviewCaptureError>) {
        if let Ok(mut current) = self.result.lock()
            && current.is_none()
        {
            *current = Some(result);
            self.ready.notify_all();
        }
    }
}

struct PendingCapture {
    connection_id: u64,
    request: PreviewCaptureRequest,
    slot: Arc<CaptureSlot>,
}

pub(crate) struct PreviewCaptureCoordinator {
    push: Arc<PushBus>,
    clients: Mutex<BTreeSet<u64>>,
    pending: Mutex<HashMap<String, PendingCapture>>,
    timeout: Duration,
}

impl PreviewCaptureCoordinator {
    pub(crate) fn new(push: Arc<PushBus>) -> Self {
        Self::with_timeout(push, DEFAULT_TIMEOUT)
    }

    fn with_timeout(push: Arc<PushBus>, timeout: Duration) -> Self {
        Self {
            push,
            clients: Mutex::new(BTreeSet::new()),
            pending: Mutex::new(HashMap::new()),
            timeout,
        }
    }

    pub(crate) fn available(&self) -> bool {
        self.clients.lock().is_ok_and(|clients| !clients.is_empty())
    }

    pub(crate) fn set_capability(
        &self,
        connection_id: u64,
        available: bool,
    ) -> Result<(), PreviewCaptureError> {
        if available {
            self.clients
                .lock()
                .map_err(|_| PreviewCaptureError::Poisoned)?
                .insert(connection_id);
        } else {
            self.remove(connection_id);
        }
        Ok(())
    }

    pub(crate) fn remove(&self, connection_id: u64) {
        if let Ok(mut clients) = self.clients.lock() {
            clients.remove(&connection_id);
        }
        let disconnected = if let Ok(mut pending) = self.pending.lock() {
            let ids = pending
                .iter()
                .filter(|(_, capture)| capture.connection_id == connection_id)
                .map(|(id, _)| id.clone())
                .collect::<Vec<_>>();
            ids.into_iter()
                .filter_map(|id| pending.remove(&id).map(|capture| capture.slot))
                .collect::<Vec<_>>()
        } else {
            Vec::new()
        };
        for slot in disconnected {
            slot.complete(Err(PreviewCaptureError::Disconnected));
        }
    }

    pub(crate) fn capture(
        &self,
        url: String,
        viewports: Vec<PreviewViewport>,
    ) -> Result<Vec<PreviewScreenshot>, PreviewCaptureError> {
        let connection_id = self
            .clients
            .lock()
            .map_err(|_| PreviewCaptureError::Poisoned)?
            .iter()
            .next()
            .copied()
            .ok_or(PreviewCaptureError::Unavailable)?;
        let request = PreviewCaptureRequest {
            request_id: Uuid::new_v4().to_string(),
            url,
            viewports,
        };
        let slot = Arc::new(CaptureSlot::new());
        self.pending
            .lock()
            .map_err(|_| PreviewCaptureError::Poisoned)?
            .insert(
                request.request_id.clone(),
                PendingCapture {
                    connection_id,
                    request: request.clone(),
                    slot: Arc::clone(&slot),
                },
            );
        if let Err(error) = self.push.send(
            connection_id,
            channel::PREVIEW_CAPTURE_REQUESTED,
            request.clone(),
        ) {
            self.pending
                .lock()
                .map_err(|_| PreviewCaptureError::Poisoned)?
                .remove(&request.request_id);
            return Err(PreviewCaptureError::Encode(error.to_string()));
        }

        let deadline = Instant::now() + self.timeout;
        let mut result = slot
            .result
            .lock()
            .map_err(|_| PreviewCaptureError::Poisoned)?;
        while result.is_none() {
            let now = Instant::now();
            if now >= deadline {
                break;
            }
            let waited = slot
                .ready
                .wait_timeout(result, deadline.saturating_duration_since(now))
                .map_err(|_| PreviewCaptureError::Poisoned)?;
            result = waited.0;
            if waited.1.timed_out() {
                break;
            }
        }
        if let Some(result) = result.take() {
            return result;
        }
        drop(result);
        if let Ok(mut pending) = self.pending.lock() {
            pending.remove(&request.request_id);
        }
        Err(PreviewCaptureError::TimedOut)
    }

    pub(crate) fn complete(
        &self,
        connection_id: u64,
        result: PreviewCaptureResult,
    ) -> Result<(), PreviewCaptureError> {
        let request_id = match &result {
            PreviewCaptureResult::Completed { request_id, .. }
            | PreviewCaptureResult::Failed { request_id, .. } => request_id,
        };
        let capture = {
            let mut pending = self
                .pending
                .lock()
                .map_err(|_| PreviewCaptureError::Poisoned)?;
            if pending
                .get(request_id)
                .is_none_or(|capture| capture.connection_id != connection_id)
            {
                return Err(PreviewCaptureError::UnknownRequest);
            }
            pending
                .remove(request_id)
                .ok_or(PreviewCaptureError::UnknownRequest)?
        };
        let completed = match result {
            PreviewCaptureResult::Failed { error, .. } => Err(PreviewCaptureError::Failed(error)),
            PreviewCaptureResult::Completed { screenshots, .. }
                if screenshots.len() == capture.request.viewports.len()
                    && screenshots.iter().zip(&capture.request.viewports).all(
                        |(screenshot, viewport)| {
                            screenshot.width == viewport.width
                                && screenshot.height == viewport.height
                        },
                    ) =>
            {
                Ok(screenshots)
            }
            PreviewCaptureResult::Completed { .. } => Err(PreviewCaptureError::UnexpectedViewports),
        };
        capture.slot.complete(completed);
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn unavailable_capture_fails_without_creating_pending_work() {
        let coordinator = PreviewCaptureCoordinator::with_timeout(
            Arc::new(PushBus::new()),
            Duration::from_millis(10),
        );
        assert_eq!(
            coordinator.capture(
                "http://127.0.0.1:5173/".into(),
                vec![PreviewViewport {
                    width: 390,
                    height: 844,
                }],
            ),
            Err(PreviewCaptureError::Unavailable)
        );
    }
}
